File: server/lib/indexing_service.rb
require_relative 'worker_pool'
require_relative 'image_comparator'

class IndexingService
  BATCH_SIZE = 100  # 每批次插入100条记录

  def initialize(project)
    @project = project
    @worker_pool = WorkerPool.new(4)  # 4个工作线程
    @source_batch = []
    @target_batches = {}  # target_id => batch array
    @mutex = Mutex.new
  end

  def process
    @project.update(status: 'indexing')

    begin
      # 索引source文件（并行处理 + 批量插入）
      index_source_files

      # 索引target文件（并行处理 + 批量插入）
      index_target_files

      @project.update(status: 'indexed')
    rescue => e
      @project.update(status: 'error', error_message: e.message)
      puts "IndexingService error: #{e.message}"
      puts e.backtrace
    end
  end

  private

  def index_source_files
    images = Dir.glob(File.join(@project.source_path, "**", "*.{jpg,jpeg,png,webp,bmp}"), File::FNM_CASEFOLD)

    puts "Found #{images.size} source images to index"

    # 批量检查已存在的文件（避免 N+1 查询）
    existing_paths = SourceFile.where(project: @project)
                               .pluck(:relative_path)
                               .to_set

    # 过滤掉已存在的文件
    new_images = images.reject do |full_path|
      relative_path = Pathname.new(full_path).relative_path_from(Pathname.new(@project.source_path)).to_s
      existing_paths.include?(relative_path)
    end

    puts "#{new_images.size} new source images to process"

    # 启动线程池
    @worker_pool.start

    # 并行处理图片
    new_images.each do |full_path|
      @worker_pool.add_job do
        process_single_source(full_path)
      end
    end

    # 等待所有任务完成
    @worker_pool.stop

    # 批量插入剩余数据
    flush_source_batch
  end

  def process_single_source(full_path)
    relative_path = Pathname.new(full_path).relative_path_from(Pathname.new(@project.source_path)).to_s

    img = nil
    comparator = nil

    begin
      # 读取图片信息
      img = Magick::Image.ping(full_path).first
      width = img.columns
      height = img.rows
      size = File.size(full_path)

      # 立即释放 ping 的图片对象
      img.destroy!
      img = nil

      # 计算哈希值
      comparator = ImageComparator.new(full_path)

      # 准备记录数据
      record = {
        project_id: @project.id,
        relative_path: relative_path,
        full_path: full_path,
        width: width,
        height: height,
        size_bytes: size,
        aspect_ratio: width.to_f / height,
        area: width * height,
        phash: comparator.phash.to_s,
        ahash: comparator.ahash.to_s,
        dhash: comparator.dhash.to_s,
        histogram: comparator.histogram.to_json,
        status: 'indexed',
        created_at: Time.now,
        updated_at: Time.now
      }

      # 线程安全地添加到批次
      @mutex.synchronize do
        @source_batch << record

        # 当批次达到阈值时，批量插入
        if @source_batch.size >= BATCH_SIZE
          flush_source_batch
        end
      end

      puts "Processed source: #{relative_path}"
    rescue => e
      puts "Error processing source #{full_path}: #{e.message}"
    ensure
      # 及时释放所有图片对象引用
      if img
        img.destroy!
        img = nil
      end
      comparator = nil
    end
  end

  def flush_source_batch
    return if @source_batch.empty?

    batch = @source_batch.dup
    @source_batch.clear

    begin
      SourceFile.insert_all(batch)
      puts "Batch inserted #{batch.size} source files"
    rescue => e
      puts "Error batch inserting source files: #{e.message}"
    end
  end

  def index_target_files
    @project.project_targets.each do |target|
      images = Dir.glob(File.join(target.path, "**", "*.{jpg,jpeg,png,webp,bmp}"), File::FNM_CASEFOLD)

      puts "Found #{images.size} target images in #{target.name}"

      # 批量检查已存在的文件
      existing_paths = TargetFile.where(project_target: target)
                                 .pluck(:relative_path)
                                 .to_set

      # 过滤掉已存在的文件
      new_images = images.reject do |full_path|
        relative_path = Pathname.new(full_path).relative_path_from(Pathname.new(target.path)).to_s
        existing_paths.include?(relative_path)
      end

      puts "#{new_images.size} new target images to process"

      # 初始化该 target 的批次数组
      @target_batches[target.id] = []

      # 启动线程池（重新启动）
      @worker_pool.start

      # 并行处理图片
      new_images.each do |full_path|
        @worker_pool.add_job do
          process_single_target(target, full_path)
        end
      end

      # 等待所有任务完成
      @worker_pool.stop

      # 批量插入剩余数据
      flush_target_batch(target.id)
    end
  end

  def process_single_target(target, full_path)
    relative_path = Pathname.new(full_path).relative_path_from(Pathname.new(target.path)).to_s

    img = nil
    comparator = nil

    begin
      # 读取图片信息
      img = Magick::Image.ping(full_path).first
      width = img.columns
      height = img.rows
      size = File.size(full_path)

      # 立即释放 ping 的图片对象
      img.destroy!
      img = nil

      # 计算哈希值
      comparator = ImageComparator.new(full_path)

      # 准备记录数据
      record = {
        project_target_id: target.id,
        relative_path: relative_path,
        full_path: full_path,
        width: width,
        height: height,
        size_bytes: size,
        aspect_ratio: width.to_f / height,
        area: width * height,
        phash: comparator.phash.to_s,
        ahash: comparator.ahash.to_s,
        dhash: comparator.dhash.to_s,
        histogram: comparator.histogram.to_json,
        created_at: Time.now,
        updated_at: Time.now
      }

      # 线程安全地添加到批次
      @mutex.synchronize do
        @target_batches[target.id] ||= []
        @target_batches[target.id] << record

        # 当批次达到阈值时，批量插入
        if @target_batches[target.id].size >= BATCH_SIZE
          flush_target_batch(target.id)
        end
      end

      puts "Processed target: #{relative_path}"
    rescue => e
      puts "Error processing target #{full_path}: #{e.message}"
    ensure
      # 及时释放所有图片对象引用
      if img
        img.destroy!
        img = nil
      end
      comparator = nil
    end
  end

  def flush_target_batch(target_id)
    return unless @target_batches[target_id]&.any?

    batch = @target_batches[target_id].dup
    @target_batches[target_id].clear

    begin
      TargetFile.insert_all(batch)
      puts "Batch inserted #{batch.size} target files for target #{target_id}"
    rescue => e
      puts "Error batch inserting target files: #{e.message}"
    end
  end
end

File: go-server/internal/services/indexing_service.go
package services

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/bilibili/look-alike/internal/database"
	"github.com/bilibili/look-alike/internal/image"
	"github.com/bilibili/look-alike/internal/models"
)

var supportedExtensions = []string{".jpg", ".jpeg", ".png", ".webp", ".bmp", ".gif", ".tiff", ".tif"}

// IndexingService handles indexing of source and target files
type IndexingService struct {
	project *models.Project
}

// NewIndexingService creates a new indexing service
func NewIndexingService(project *models.Project) *IndexingService {
	return &IndexingService{
		project: project,
	}
}

// Process runs the indexing process
func Process(project *models.Project) error {
	log.Printf("========================================")
	log.Printf("IndexingService.Process started")
	log.Printf("Project: %s (ID: %d)", project.Name, project.ID)
	log.Printf("Source path: %s", project.SourcePath)
	log.Printf("========================================")

	// Update project status
	if err := database.DB.Model(&project).Update("status", "indexing").Error; err != nil {
		return err
	}

	svc := NewIndexingService(project)

	// Index source files
	if err := svc.indexSourceFiles(); err != nil {
		database.DB.Model(&project).Updates(map[string]interface{}{
			"status":        "error",
			"error_message": err.Error(),
		})
		return fmt.Errorf("failed to index source files: %w", err)
	}

	// Index target files
	if err := svc.indexTargetFiles(); err != nil {
		database.DB.Model(&project).Updates(map[string]interface{}{
			"status":        "error",
			"error_message": err.Error(),
		})
		return fmt.Errorf("failed to index target files: %w", err)
	}

	// Update project status to indexed
	if err := database.DB.Model(&project).Update("status", "indexed").Error; err != nil {
		return err
	}

	log.Println("[SUCCESS] IndexingService completed")
	log.Println("========================================")
	return nil
}

// indexSourceFiles indexes all source files
func (svc *IndexingService) indexSourceFiles() error {
	log.Printf("[SOURCE] Scanning source directory: %s", svc.project.SourcePath)

	sourcePath := strings.TrimSpace(svc.project.SourcePath)
	if _, err := os.Stat(sourcePath); os.IsNotExist(err) {
		return fmt.Errorf("source path does not exist: %s", sourcePath)
	}

	// Find all image files
	images, err := findImages(sourcePath)
	if err != nil {
		return err
	}

	log.Printf("[SOURCE] Found %d source images", len(images))

	if len(images) == 0 {
		return fmt.Errorf("no source images found in: %s", sourcePath)
	}

	// Get existing files to avoid duplicates
	var existingFiles []models.SourceFile
	database.DB.Where("project_id = ?", svc.project.ID).Select("relative_path").Find(&existingFiles)
	existingPaths := make(map[string]bool)
	for _, f := range existingFiles {
		existingPaths[f.RelativePath] = true
	}

	// Filter new images
	var newImages []string
	for _, imgPath := range images {
		relPath, _ := filepath.Rel(sourcePath, imgPath)
		if !existingPaths[relPath] {
			newImages = append(newImages, imgPath)
		}
	}

	log.Printf("[SOURCE] %d new source images to process", len(newImages))

	// Process images concurrently
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, 4) // Limit to 4 concurrent goroutines
	var mu sync.Mutex
	var batch []models.SourceFile
	const batchSize = 100

	for _, imgPath := range newImages {
		wg.Add(1)
		go func(path string) {
			defer wg.Done()
			semaphore <- struct{}{} // Acquire
			defer func() { <-semaphore }() // Release

			sourceFile, err := processSourceFile(path, sourcePath, svc.project.ID)
			if err != nil {
				log.Printf("[ERROR] Failed to process source file %s: %v", path, err)
				return
			}

			mu.Lock()
			batch = append(batch, *sourceFile)
			if len(batch) >= batchSize {
				if err := database.DB.Create(&batch).Error; err != nil {
					log.Printf("[ERROR] Failed to batch insert source files: %v", err)
				} else {
					log.Printf("[SOURCE] Batch inserted %d source files", len(batch))
				}
				batch = nil
			}
			mu.Unlock()
		}(imgPath)
	}

	wg.Wait()

	// Insert remaining batch
	if len(batch) > 0 {
		if err := database.DB.Create(&batch).Error; err != nil {
			return fmt.Errorf("failed to insert remaining source files: %w", err)
		}
		log.Printf("[SOURCE] Inserted final batch of %d source files", len(batch))
	}

	log.Println("[SOURCE] Source file indexing completed")
	return nil
}

// indexTargetFiles indexes all target files
func (svc *IndexingService) indexTargetFiles() error {
	var targets []models.ProjectTarget
	if err := database.DB.Where("project_id = ?", svc.project.ID).Find(&targets).Error; err != nil {
		return err
	}

	log.Printf("[TARGET] Indexing %d target directories", len(targets))

	for _, target := range targets {
		if err := svc.indexSingleTarget(&target); err != nil {
			log.Printf("[ERROR] Failed to index target %s: %v", target.Name, err)
			continue
		}
	}

	log.Println("[TARGET] Target file indexing completed")
	return nil
}

// indexSingleTarget indexes files for a single target
func (svc *IndexingService) indexSingleTarget(target *models.ProjectTarget) error {
	log.Printf("[TARGET] Indexing target: %s (%s)", target.Name, target.Path)

	targetPath := strings.TrimSpace(target.Path)
	if _, err := os.Stat(targetPath); os.IsNotExist(err) {
		return fmt.Errorf("target path does not exist: %s", targetPath)
	}

	images, err := findImages(targetPath)
	if err != nil {
		return err
	}

	log.Printf("[TARGET] Found %d images in target %s", len(images), target.Name)

	// Get existing files
	var existingFiles []models.TargetFile
	database.DB.Where("project_target_id = ?", target.ID).Select("relative_path").Find(&existingFiles)
	existingPaths := make(map[string]bool)
	for _, f := range existingFiles {
		existingPaths[f.RelativePath] = true
	}

	// Filter new images
	var newImages []string
	for _, imgPath := range images {
		relPath, _ := filepath.Rel(targetPath, imgPath)
		if !existingPaths[relPath] {
			newImages = append(newImages, imgPath)
		}
	}

	log.Printf("[TARGET] %d new images to process for target %s", len(newImages), target.Name)

	// Process concurrently
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, 4)
	var mu sync.Mutex
	var batch []models.TargetFile
	const batchSize = 100

	for _, imgPath := range newImages {
		wg.Add(1)
		go func(path string) {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()

			targetFile, err := processTargetFile(path, targetPath, target.ID)
			if err != nil {
				log.Printf("[ERROR] Failed to process target file %s: %v", path, err)
				return
			}

			mu.Lock()
			batch = append(batch, *targetFile)
			if len(batch) >= batchSize {
				if err := database.DB.Create(&batch).Error; err != nil {
					log.Printf("[ERROR] Failed to batch insert target files: %v", err)
				} else {
					log.Printf("[TARGET] Batch inserted %d target files", len(batch))
				}
				batch = nil
			}
			mu.Unlock()
		}(imgPath)
	}

	wg.Wait()

	// Insert remaining batch
	if len(batch) > 0 {
		if err := database.DB.Create(&batch).Error; err != nil {
			return fmt.Errorf("failed to insert remaining target files: %w", err)
		}
		log.Printf("[TARGET] Inserted final batch of %d target files", len(batch))
	}

	return nil
}

// processSourceFile processes a single source file
func processSourceFile(fullPath, basePath string, projectID uint) (*models.SourceFile, error) {
	relPath, err := filepath.Rel(basePath, fullPath)
	if err != nil {
		return nil, err
	}

	// Get file info
	fileInfo, err := os.Stat(fullPath)
	if err != nil {
		return nil, err
	}

	// Calculate phash and color histogram
	comparator, err := image.NewImageComparator(fullPath)
	if err != nil {
		return nil, err
	}

	// Convert color histogram to JSON
	histogramJSON, err := json.Marshal(comparator.ColorHistogram)
	if err != nil {
		return nil, err
	}

	// Calculate aspect ratio and area
	aspectRatio := float64(comparator.Width) / float64(comparator.Height)
	area := comparator.Width * comparator.Height

	sourceFile := &models.SourceFile{
		ProjectID:    projectID,
		RelativePath: relPath,
		FullPath:     fullPath,
		Width:        comparator.Width,
		Height:       comparator.Height,
		SizeBytes:    fileInfo.Size(),
		Status:       "indexed",
		AspectRatio:  aspectRatio,
		Area:         area,
		Phash:        fmt.Sprintf("%d", comparator.Phash),
		Ahash:        "", // Not used
		Dhash:        "", // Not used
		Histogram:    string(histogramJSON),
	}

	return sourceFile, nil
}

// processTargetFile processes a single target file
func processTargetFile(fullPath, basePath string, targetID uint) (*models.TargetFile, error) {
	relPath, err := filepath.Rel(basePath, fullPath)
	if err != nil {
		return nil, err
	}

	fileInfo, err := os.Stat(fullPath)
	if err != nil {
		return nil, err
	}

	// Calculate phash and color histogram
	comparator, err := image.NewImageComparator(fullPath)
	if err != nil {
		return nil, err
	}

	// Convert color histogram to JSON
	histogramJSON, err := json.Marshal(comparator.ColorHistogram)
	if err != nil {
		return nil, err
	}

	aspectRatio := float64(comparator.Width) / float64(comparator.Height)
	area := comparator.Width * comparator.Height

	targetFile := &models.TargetFile{
		ProjectTargetID: targetID,
		FullPath:        fullPath,
		RelativePath:    relPath,
		Width:           comparator.Width,
		Height:          comparator.Height,
		SizeBytes:       fileInfo.Size(),
		AspectRatio:     aspectRatio,
		Area:            area,
		Phash:           fmt.Sprintf("%d", comparator.Phash),
		Ahash:           "", // Not used
		Dhash:           "", // Not used
		Histogram:       string(histogramJSON),
	}

	return targetFile, nil
}

// findImages finds all image files in a directory recursively
func findImages(rootPath string) ([]string, error) {
	var images []string

	err := filepath.Walk(rootPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if info.IsDir() {
			return nil
		}

		ext := strings.ToLower(filepath.Ext(path))
		for _, supportedExt := range supportedExtensions {
			if ext == supportedExt {
				images = append(images, path)
				break
			}
		}

		return nil
	})

	return images, err
}
